File: backend/core/generico.py
# -*- coding: utf-8 -*-
"""Si un alimento es GENERICO o es DE MARCA. Una sola definicion, la misma que el front.

El catalogo no guarda el dato: `db.foods` tiene {nombre, categorias, racion, macros, unidades,
minimo, url, imagen, codigoBarras} y ninguno de `marca`, `es_generico` o `sin_web`
(comprobado en produccion el 29-08-2026). Se deduce, y se deducia solo de la URL.

Francisco, 29-08-2026: «el filtro de generico no esta funcionando bien, se le escapan
alimentos con marca, no pasa en todas las categorias». En produccion hay SEIS alimentos de
marca sin enlace, y con «sin enlace = generico» se colaban todos:

    Chicharron iberico (7 Hermanos)          Batido proteico Smart Protein... (Nutrisport)
    Levadura nutricional (Sol Natural)       Cafe con leche sin azucares... (Hacendado)
    Barrita proteica apple pie (Nutrisport)  Copos de avena integral sin gluten (Esgir)

Se miran las dos cosas: el enlace y el parentesis del nombre (el criterio que confirmo
Francisco el 08-07: 2.736 de 3.211 lo llevan). El parentesis solo no basta porque a veces es
una ACLARACION. Se listaron los 18 que hay entre los 475 sin enlace y salen 12 aclaraciones
-- tres patrones -- y 6 marcas; de ahi la lista de abajo.

Lo que NO cubre: una marca sin parentesis y sin enlace (Aquarius, Gatorade, Monster son asi,
aunque hoy los tres tienen enlace). Para eso hace falta un campo en el catalogo.

El gemelo en el front es `frontend/src/lib/generico.js`: si se toca uno, se toca el otro.
"""
import re

# Parentesis que aclaran el alimento en vez de nombrar una marca. Salen del catalogo real.
ACLARACIONES = re.compile(
    r"corte magro|corte graso|m[áa]s del|menos del|\d\s*%|c[áa]scara|al natural|en conserva|"
    r"congelad|cocinad|cocid|crud|desgrasad|sin az[úu]car|sin sal|light|enter[oa]|peque[nñ]|"
    r"grande|mediano|por gramos|macros orientativos|aprox",
    re.I,
)

_PARENTESIS = re.compile(r"\(([^)]+)\)")


def marca_del_nombre(nombre):
    """El texto del parentesis si parece una MARCA; None si es aclaracion o no hay."""
    m = _PARENTESIS.search(nombre or "")
    if not m:
        return None
    dentro = m.group(1).strip()
    return None if ACLARACIONES.search(dentro) else dentro


def es_de_marca(food):
    """De marca: tiene enlace a la web del super, o lleva la marca en el nombre."""
    food = food or {}
    return bool(food.get("url")) or bool(marca_del_nombre(food.get("nombre")))


def es_generico(food):
    """Generico: ni enlace ni marca en el nombre."""
    return not es_de_marca(food)

File: frontend/src/lib/generico.js
/**
 * ¿ESTE ALIMENTO ES GENÉRICO O ES DE MARCA? EN UN SOLO SITIO.
 *
 * El catálogo NO guarda el dato: `db.foods` tiene `{nombre, categorias, racion, macros,
 * unidades, minimo, url, imagen, codigoBarras}` y nada más -- comprobado en producción, no
 * hay `marca`, ni `es_generico`, ni `sin_web`. Así que se deduce, y hasta hoy se deducía de
 * la URL: «sin enlace = genérico».
 *
 * Francisco, 29-08-2026: «el filtro de genérico no está funcionando bien, se le escapan
 * alimentos con marca, no pasa en todas las categorías». Y era eso: en producción hay SEIS
 * alimentos de marca sin enlace, y el filtro los daba por genéricos. Cada uno en una
 * categoría distinta -- 29, 51, 2.1, 5.4, 19.3.2 y 7.1.2 --, de ahí lo de «no en todas»:
 *
 *   Chicharrón ibérico (7 Hermanos)          Batido proteico Smart Protein... (Nutrisport)
 *   Levadura nutricional (Sol Natural)       Café con leche sin azúcares... (Hacendado)
 *   Barrita proteica apple pie (Nutrisport)  Copos de avena integral sin gluten (Esgir)
 *
 * SE MIRAN LAS DOS COSAS: el enlace y el paréntesis del nombre. El paréntesis es el criterio
 * que confirmó Francisco el 08-07 (2.736 de 3.211 lo llevan), pero él solo no vale porque hay
 * paréntesis que son ACLARACIÓN y no marca. Se listaron todos los que hay entre los 475 sin
 * enlace: son 18 y se parten limpios en 12 aclaraciones (tres patrones nada más) y 6 marcas.
 * De ahí sale ACLARACIONES: si el paréntesis encaja ahí, sigue siendo genérico.
 *
 * Lo que esto NO resuelve, y conviene saberlo: una marca sin paréntesis y sin enlace pasaría
 * igual (Aquarius, Gatorade y Monster son así, aunque hoy los tres tienen enlace). La
 * solución de verdad es un campo en el catálogo; mientras no exista, esto es lo más cerca que
 * se puede estar sin tocar 3.211 fichas.
 */

// Los paréntesis que son una aclaración del alimento, no una marca. Salen de mirar los 18
// que hay en el catálogo entre los alimentos sin enlace, no de imaginarlos.
const ACLARACIONES = /corte magro|corte graso|m[áa]s del|menos del|\d\s*%|c[áa]scara|al natural|en conserva|congelad|cocinad|cocid|crud|desgrasad|sin az[úu]car|sin sal|light|enter[oa]|peque[nñ]|grande|mediano|por gramos|macros orientativos|aprox/i;

/** El texto del paréntesis si parece una MARCA; null si es una aclaración o no hay. */
export const marcaDelNombre = (nombre) => {
    const m = /\(([^)]+)\)/.exec(nombre || '');
    if (!m) return null;
    const dentro = m[1].trim();
    return ACLARACIONES.test(dentro) ? null : dentro;
};

/** De marca: tiene enlace a la web del súper, o lleva la marca en el nombre. */
export const esDeMarca = (food) => Boolean(food?.url) || Boolean(marcaDelNombre(food?.nombre));

/** Genérico: ni enlace ni marca en el nombre. */
export const esGenerico = (food) => !esDeMarca(food);

export default esGenerico;
